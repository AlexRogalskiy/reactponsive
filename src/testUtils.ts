export const generateMatchMediaMock = () => {
  const isMatching = (query: string) =>
    query === '(min-width: 400px)' || query === '(min-width: 1024px)';

  window.matchMedia = jest.fn().mockImplementation(query => {
    return {
      matches: isMatching(query),
      media: query,
      onchange: null,
      addListener: jest.fn(x =>
        x({
          media: query,
          matches: isMatching(query),
        }),
      ),
      removeListener: jest.fn(),
    };
  });
};
