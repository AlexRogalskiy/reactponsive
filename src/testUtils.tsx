import { render } from '@testing-library/react';
import { renderHook, RenderHookOptions } from '@testing-library/react-hooks';
import React from 'react';

import Provider from './components/Provider';
import { IHasAliases } from './types';

type F<MqStr> = (query: MqStr) => MediaQueryListEvent;

export const TABLET = '(min-width: 768px)';
export const DESKTOP = '(min-width: 1024px)';

const defaultAlias = {
  tablet: TABLET,
  desktop: DESKTOP
};

export const generateMatchMediaMock = <
  MqStr extends string,
  Value extends boolean
>(
  initialState?: Record<MqStr, Value>
) => {
  const values = new Map(
    Object.entries(
      initialState || {
        [TABLET]: true,
        [DESKTOP]: false
      }
    )
  );
  const listeners = new Map();

  const updateBreakpoint = (key: MqStr, value: Value) => {
    values.set(key, value);
    const fn = listeners.get(key);
    if (fn) {
      fn({
        media: key,
        matches: value
      });
    }
  };

  window.matchMedia = jest.fn().mockImplementation((query: MqStr) => ({
    matches: values.get(query) === true,
    media: query,
    onchange: null,
    addListener: (fn: F<MqStr>) => listeners.set(query, fn),
    removeListener: jest.fn(() => listeners.delete(query))
  }));

  return {
    updateBreakpoint
  };
};

export const renderWithProvider = (
  ui: any,
  { alias, ...options }: Partial<IHasAliases> = {
    alias: defaultAlias
  }
) =>
  render(ui, {
    ...options,
    // @ts-ignore
    wrapper: ({ children }) => <Provider alias={alias}>{children}</Provider>
  });

export const renderHookWithProvider = (
  callback: (props?: unknown) => unknown,
  {
    alias,
    ...options
  }: Partial<IHasAliases> & {
    options?: RenderHookOptions<{}>;
  } = {
    alias: defaultAlias
  }
) =>
  renderHook(callback, {
    ...options,
    // @ts-ignore
    wrapper: ({ children }) => <Provider alias={alias}>{children}</Provider>
  });
