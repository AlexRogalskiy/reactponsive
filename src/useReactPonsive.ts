import { useState, useEffect } from 'react';

import { Mqs } from './types';
import { buildMqObj, getReactPonsiveProps, updateMqsMatches } from './_lib';

const useReactPonsive = (mqsStringArr: Mqs) => {
  const initialMqs = mqsStringArr.map(buildMqObj);
  const [mqs, setMqs] = useState(initialMqs);
  useEffect(() => {
    const updateMatches = (e: MediaQueryListEvent) => {
      const newMqs = updateMqsMatches(mqs, e);
      setMqs(newMqs);
    }
    mqs.forEach(mqObj => {
      mqObj.mq.addListener(updateMatches);
    });

    return () => {
      mqs.forEach(mqObj => {
        mqObj.mq.removeListener(updateMatches);
      });
    };
  }, [mqs, setMqs]);

  const props = getReactPonsiveProps(mqs);
  return props;
}

export default useReactPonsive;