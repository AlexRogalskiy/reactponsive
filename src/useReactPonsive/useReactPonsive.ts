import { useState, useEffect } from 'react';

import { Mqs } from '../types';
import { buildMqObj, buildReactPonsiveProps, updateMqsMatches } from '../utils';
import useAlias from '../useAlias';

const useReactPonsive = (mqsStringArr: Mqs) => {
  const alias = useAlias();
  const buildMqObjWithAlias = buildMqObj(alias);
  const initialMqs = mqsStringArr.map(buildMqObjWithAlias);
  const [mqs, setMqs] = useState(initialMqs);
  const updateMatches = (e: MediaQueryListEvent) => {
    const newMqs = updateMqsMatches(mqs, e);
    setMqs(newMqs);
  };
  useEffect(() => {
    const mqEvents = mqs.map(({ value }) => {
      const mq = window.matchMedia(value);
      mq.addListener(updateMatches);
      return mq;
    });
    return () => {
      mqEvents.forEach(mq => mq.removeListener(updateMatches));
    };
  }, [mqsStringArr]);
  return buildReactPonsiveProps(mqs);
};

export default useReactPonsive;
