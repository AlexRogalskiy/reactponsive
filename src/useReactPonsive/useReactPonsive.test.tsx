import * as React from 'react';
import { render } from 'react-testing-library';
import 'jest-dom/extend-expect';

import useReactPonsive from './useReactPonsive';
import { generateMatchMediaMock } from '../testUtils';

generateMatchMediaMock();

const MyComp = () => {
  const rpValues = useReactPonsive([
    '(min-width: 400px)',
    '(min-width: 768px)',
    '(min-width: 1024px)',
  ]);
  return <div data-testid="reactponsive" {...rpValues} />;
};

describe('useReactPonsive', () => {
  test('it works', () => {
    const element = <MyComp />;
    const { getByTestId } = render(element);
    expect(getByTestId('reactponsive').attributes).toMatchSnapshot();
  });
});
