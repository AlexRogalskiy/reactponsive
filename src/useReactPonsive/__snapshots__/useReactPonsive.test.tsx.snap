// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useReactPonsive it works 1`] = `
NamedNodeMap {
  "data-testid": "reactponsive",
  "first": "(min-width: 400px)",
  "last": "(min-width: 1024px)",
  "matches": "[object Object]",
  "passes": "(min-width: 400px),(min-width: 1024px)",
}
`;
